### PR TITLE
Fix Archetypes reference migration

### PR DESCRIPTION
**Pull request by rpatterson**

### CODE REVIEW

**Comment by rpatterson** [comment]
Required for https://github.com/plone/plone.app.contenttypes/issues/510
@jenkins-plone-org please run jobs

**Comment by rpatterson** [comment]
Fixes #22

**Comment by rpatterson** [comment]
@jenkins-plone-org please run jobs

**Comment by rpatterson** [comment]
@jenkins-plone-org please run jobs

**Comment by pbauer** [comment]
I cannot really say if that is ok. I'm worried that there might be a side-effect to not stting the childen into `self._objects`. Skipping reindexing should be ok thought. 
If you are sure that migrations still work with this change I have no objections. You should add the same change to the master-branch as well though.

File: Products/contentmigration/basemigrator/migrator.py [hunk 1]
```diff
@@ -574,23 +573,13 @@ def migrate_children(self):
...
-                # this.
-                if id in self.new.objectIds():
-                    self.new._delOb(id)
-                    if getattr(self.new, '_objects', None) is not None:
-                        self.new._objects = tuple([
-                            o for o in self.new._objects if o['id'] != id])
-                    self.new._setOb(id, obj)
-                else:
-                    raise
+            if id in self.new.objectIds():
```
Comment (gforcada) [inline Products/contentmigration/basemigrator/migrator.py]: please avoid using python builtins, `object_id` would be as fine as well :wink:

**Comment by pbauer** [comment]
I adapted these changes to the master-branch with https://github.com/plone/Products.contentmigration/pull/24 and the test `test_default_pages_are_kept_during_migration` is failing. 

@rpatterson The reason is your reimplementation of `migrate_children` which no longer stores the objects in `self.subobjs` and sets them again with `self.new._setObject()`. 

Can you comment why you chose to use `utils.unrestricted_move` instead? That seems to be less efficient since it triggers a lot of subscribers for the subobjects. And also for some reason the default_page is lost.

There are a lot of things here that I like but that change looks wrong to me.

**Comment by dest81** [comment]
@pbauer 
I am testing this changes now. I found only two bugs:
 - as you mentioned before default page is lost
 - children's order in the new folder is wrong 

Here is the commit to fix this two bugs https://github.com/dest81/Products.contentmigration/commit/4b316ac2fb97c5965ecea0ff393a06bcf73fc67e

I can send PR to apply_rpatterson_changes branch if you agree with it.

Let you know if I find more bugs

**Comment by pbauer** [comment]
@dest81 yes, please create a pull-request

**Comment by pbauer** [comment]
I finally looked at https://github.com/plone/Products.contentmigration/pull/23/commits/aa89e57243824b734204b4f64536e12c19f47195 (especially the commit-message) more closely and I now think that the change is ok and can be merged once the fixes by @dest81 are included.

But: For all of my use-cases the old optimized version also worked and I have not compared the performance of the two implementations.

**Comment by pbauer** [comment]
@jenkins-plone-org please run jobs

**Comment by jensens** [comment]
There was no activity on this for a while. I close this request for now. If you plan to work further on this please reopen and go on.

**Comment by rpatterson** [comment]
> @rpatterson The reason is your reimplementation of `migrate_children` which no longer stores the objects in `self.subobjs` and sets them again with `self.new._setObject()`.
> 
> Can you comment why you chose to use `utils.unrestricted_move` instead? That seems to be less efficient since it triggers a lot of subscribers for the subobjects. And also for some reason the default_page is lost.

My apologies for not capturing the why for this and other changes in my commits, or at the very least in this PR's description.  I'm much better about that these days, not that it helps here.

At any rate, I dimly recall there being a cascade of problems in real-world, client data that arose from these optimizations.  IOW, there were real consequences to *not* running all those subscribers that broke things in real-world data.  But that's all I recall at this point so long after I made the changes.  Again, sorry.

**Pull request closed without merging**